### PR TITLE
[WIP] add tslint type check test

### PR DESCRIPTION
**Pull request by cncolder**
try refactoring list type definition

This pr use for discussion. Working in progress. **DONOT** merge.

ref #1557 
ref ant-design/antd-mobile-samples#21

After this pr. React Native user need update `tsconfig.json`

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "antd-mobile": [
        "./node_modules/antd-mobile/lib/native"
      ]
    }
  }
}
```

Once you finish configure path mapping. You will get nearly perfect vscode IntelliSense for List component.

First of all, thank you for your contribution! :-)

Please makes sure that these checkboxes are checked before submitting your PR, thank you!

* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [ ] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [x] Rebase before creating a PR to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you PR.

Extra checklist:

**if** *isBugFix* **:**

  * [ ] Make sure that you add at least one unit test for the bug which you had fixed.

**elif** *isNewFeature* **:**

  * [ ] Update API docs for the component.
  * [ ] Update/Add demo to demonstrate new feature.
  * [x] Update TypeScript definition for the component.
  * [x] Add unit tests for the feature.

---
This change is [Reviewable](https://reviewable.io/reviews/ant-design/ant-design-mobile/1560)

### CODE REVIEW

File: components/list/PropsType.tsx [hunk 1]
```diff
@@ -1,74 +1,77 @@
-import React, { ReactNode } from 'react';
+import React from 'react';
+import ReactNative from 'react-native';
```
Comment (paranoidjk) [inline components/list/PropsType.tsx]: 这里还是会导致一个目前已知的问题
- 使用 web 组件的用户需要安装 react-native 依赖

File: tests/types/web/web.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import List from 'antd-mobile/lib/list';
```
Comment (paranoidjk) [inline tests/types/web/web.tsx]: 按你这个改法，这一行 https://github.com/ant-design/ant-design-mobile/blob/feature-2.0/components/index.tsx#L1 的注释可以去掉了，umd dist 打包的入口和 types 文件生成的入口已经拆分开了

File: tsconfig.json [hunk 3]
```diff
@@ -11,7 +11,8 @@
   "exclude": [
     "node_modules",
     "lib",
-    "es"
+    "es",
```
Comment (paranoidjk) [inline tsconfig.json]: 还要改 `Pkg.typings` 吧？默认指向 web 的类型文件，rn的用户就按照 https://github.com/ant-design/antd-mobile-samples/pull/21
自定义 tsconfig 指向 native 的类型文件

File: components/list/PropsType.tsx (hunk 1 above)
Comment (cncolder) [inline components/list/PropsType.tsx]: 这个很好解决，WebProps 和 NativeProps 移到各自的组件里。

类型定义在需要的地方没什么不可以的，因为我们本来就是个 typescript 项目。 `PropsType.tsx` 存在的意义是重用代码，而且 PropsType 这个文件名也待考量，我们使用 typescript 的目的不能只局限于 React props type 验证这个低起点上。

File: tests/types/web/web.tsx (hunk 2 above)
Comment (cncolder) [inline tests/types/web/web.tsx]: 这个不是改动，是因为我只试着重写了 List 的类型定义，全部导入的话测试通不过，这个文件是为了 `tslint --type-check` 用的。

用户还是要 `import {} from 'antd-mobile'`

**Comment by paranoidjk** [comment]
@cncolder 
分支已经切过来了。

https://github.com/ant-design/ant-design-mobile/tree/master -> 2.x
https://github.com/ant-design/ant-design-mobile/tree/1.x -> 1.x

直接往 master 合吧。

**Pull request closed without merging**